File: src/main/java/se/kth/spork/spoon/StructuralConflict.java
package se.kth.spork.spoon;

import se.kth.spork.base3dm.Pcs;
import se.kth.spork.spoon.wrappers.SpoonNode;
import spoon.reflect.declaration.CtElement;

import java.util.*;

/**
 * A simple class that provides some information on a structural conflict. Meant to be put as metadata on a conflict
 * "dummy node" using {@link StructuralConflict#METADATA_KEY} as the key.
 *
 * @author Simon Larsén
 */
public class StructuralConflict {
    public static final String METADATA_KEY = "SPORK_STRUCTURAL_CONFLICT";
    public final List<CtElement> left;
    public final List<CtElement> right;
    public final Optional<List<CtElement>> base;
    public final Optional<String> lineBasedMerge;

    /**
     * @param left  The left part of the conflict.
     * @param right The right part of the conflict.
     */
    public StructuralConflict(List<CtElement> left, List<CtElement> right) {
        this.left = left;
        this.right = right;
        base = Optional.empty();
        lineBasedMerge = Optional.empty();
    }

    /**
     * Create an approximated conflict.
     */
    public StructuralConflict(CtElement base, CtElement left, CtElement right, String lineBasedMerge) {
        this.base = Optional.of(Collections.singletonList(base));
        this.left = Collections.singletonList(left);
        this.right = Collections.singletonList(right);
        this.lineBasedMerge = Optional.of(lineBasedMerge);
    }

    public Optional<String> lineBasedMerge() {
        return lineBasedMerge;
    }

    public static boolean isRootConflict(Pcs<?> left, Pcs<?> right) {
        return !Objects.equals(left.getRoot(), right.getRoot()) &&
                (Objects.equals(left.getPredecessor(), right.getPredecessor()) ||
                        Objects.equals(left.getSuccessor(), right.getSuccessor()));
    }

    public static boolean isPredecessorConflict(Pcs<?> left, Pcs<?> right) {
        return !Objects.equals(left.getPredecessor(), right.getPredecessor()) &&
                Objects.equals(left.getSuccessor(), right.getSuccessor()) &&
                Objects.equals(left.getRoot(), right.getRoot());
    }

    public static boolean isSuccessorConflict(Pcs<?> left, Pcs<?> right) {
        return !Objects.equals(left.getSuccessor(), right.getSuccessor()) &&
                Objects.equals(left.getPredecessor(), right.getPredecessor()) &&
                Objects.equals(left.getRoot(), right.getRoot());
    }

    public static Set<SpoonNode>
    extractRootConflictingNodes(Map<Pcs<SpoonNode>, Set<Pcs<SpoonNode>>> structuralConflicts) {
        Set<SpoonNode> toIgnore = new HashSet<>();

        for (Map.Entry<Pcs<SpoonNode>, Set<Pcs<SpoonNode>>> entry : structuralConflicts.entrySet()) {
            Pcs<SpoonNode> pcs = entry.getKey();
            for (Pcs<SpoonNode> other : entry.getValue()) {
                if (isRootConflict(pcs, other)) {
                    if (pcs.getPredecessor().equals(other.getPredecessor())) {
                        toIgnore.add(other.getPredecessor());
                    }
                    if (pcs.getSuccessor().equals(other.getSuccessor())) {
                        toIgnore.add(other.getSuccessor());
                    }
                }
            }
        }
        return toIgnore;
    }
}

File: src/main/kotlin/se/kth/spork/spoon/conflict/StructuralConflict.kt
package se.kth.spork.spoon.conflict

import se.kth.spork.spoon.wrappers.SpoonNode
import spoon.reflect.declaration.CtElement
import java.util.HashSet
import se.kth.spork.base3dm.Pcs

/**
 * A simple class that provides some information on a structural conflict. Meant to be put as
 * metadata on a conflict "dummy node" using [StructuralConflict.METADATA_KEY] as the key.
 *
 * @author Simon Larsén
 */
class StructuralConflict {
    val left: List<CtElement>
    val right: List<CtElement>
    val base: List<CtElement>?
    val lineBasedMerge: String?

    /**
     * @param left The left part of the conflict.
     * @param right The right part of the conflict.
     */
    constructor(left: List<CtElement>, right: List<CtElement>) {
        this.left = left
        this.right = right
        base = null
        lineBasedMerge = null
    }

    /** Create an approximated conflict.  */
    constructor(
        base: CtElement?, left: CtElement, right: CtElement, lineBasedMerge: String
    ) {
        this.base = if (base != null) listOf(base) else null
        this.left = listOf(left)
        this.right = listOf(right)
        this.lineBasedMerge = lineBasedMerge
    }

    companion object {
        const val METADATA_KEY = "SPORK_STRUCTURAL_CONFLICT"
        fun isRootConflict(left: Pcs<*>, right: Pcs<*>): Boolean {
            return (left.root != right.root
                    && (left.predecessor == right.predecessor
                    || left.successor == right.successor))
        }

        fun isPredecessorConflict(left: Pcs<*>, right: Pcs<*>): Boolean {
            return (left.predecessor != right.predecessor
                    && left.successor == right.successor
                    && left.root == right.root)
        }

        fun isSuccessorConflict(left: Pcs<*>, right: Pcs<*>): Boolean {
            return (left.successor != right.successor
                    && left.predecessor == right.predecessor
                    && left.root == right.root)
        }

        fun extractRootConflictingNodes(
            structuralConflicts: Map<Pcs<SpoonNode>, Set<Pcs<SpoonNode>>>
        ): Set<SpoonNode> {
            val toIgnore: MutableSet<SpoonNode> = HashSet()
            for ((pcs, value) in structuralConflicts) {
                for (other in value) {
                    if (isRootConflict(pcs, other)) {
                        if (pcs.predecessor == other.predecessor) {
                            toIgnore.add(other.predecessor)
                        }
                        if (pcs.successor == other.successor) {
                            toIgnore.add(other.successor)
                        }
                    }
                }
            }
            return toIgnore
        }
    }
}